open_spiel/python/algorithms/tabular_maqlearner.py

```diff
@@ -0,0 +1,209 @@
+# Copyright 2019 DeepMind Technologies Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tabular Multiagent Q-learning agent."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import abc
+import collections
+import itertools
+import numpy as np
+
+from open_spiel.python import rl_agent
+from open_spiel.python import rl_tools
+from open_spiel.python.algorithms.jpsro import _mgce, _mgcce
+from open_spiel.python.examples.matrix_nash import lemke_howson_solve
+
+def valuedict():
+  return collections.defaultdict(float)
+
+
+class JointActionSolver:
+  @abc.abstractmethod
+  def __call__(self, q_tables, action_dims):
+    '''
+      Args:
+        q_tables: a list of Q tables for each agent
```

Comment:
What is the type of the q table?

Any reason to not have not have these be 3-dimensional (player, row, col) numpy array where the indices correspond to indices in each player's legal moves list? That's how we normally do it elsewhere, and we would not have to pass in action_dimcs and it's generally easier to work with numpy arrays

Reply:
yes it makes sense. Now I just directly make this action_dims as an explicit input argument for the MAQagent (num_of_actions, which is a list now) so a JointActionSolver won't need it